text/0143-versioning-and-stability.md

```diff
@@ -0,0 +1,128 @@
+
+![drawing](img/0143_api_lifecycle.png)
+
+### Lifecycle stages
+
+**Experimental –** Bugs, breaking changes and performance issues may occur.
+
+**Stable –**  Stability guarantees apply, based on component tpye (API, SDK, Conventions, and Contrib). Long term dependencies may now be taken against these packages.
+
+**Deprecated –** this signal has been replaced but is still supported under LTS.
+
+**Removed -** a deprecated signal is no longer supported, and is removed.
+
+All signal components may become stable together, or one by one in the following order: API, SDK, Semantic Conventions, Contrib.
+
+### Stability
+
+Once a signal component is marked as stable, the following rules apply until the end of that signal’s existence.
+
+**API Stability -**
+No backward-incompatible changes to the API are allowed unless the major version number is incremented. All existing API calls must continue to compile and function against all future minor versions of the same major version. ABI compatibility is offered in languages which require it.
+
+**SDK Stability -**
+Public portions of the SDK (constructors, configuration, und-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break; ABI compatibility is not required.
+
+**Semantic Conventions Stability -**
+Semantic Conventions may not be removed once they are stable. New conventions may be added to replace usage of older conventions, but the older conventions are never removed.
+
+**Contrib Stability -**
+Plugins and instrumentation are kept up to date, and are released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of opentelemetry, and not be held back by the plugins that they depend on.
+
+Public portions of contrib packages (constructors, configuration, interfaces) must remain backwards compatible. ABI compatibility not required.
+
+Telemetry produced by contrib instrumentation must also remain stable and backwards compatible, to avoid breaking alerts and dashboard.
+
+### Deprecation
+
+In theory, signals could be replaced. When this happens, they are marked as deprecated.
+
+Code is only marked as deprecated when the replacement becomes stable. Deprecated code still abides by the same support guarantees as stable code. API’s remain stable and backwards compatible.
+
+### Removal
+
+Packages are end-of-life’d by being removed, along with a major version bump.
+
+We currently have no plans for deprecating signals or creating a major version past v1.0.
+
+## Version Numbers
+
+OpenTelemetry follows [semver 2.0](https://semver.org/) conventions, with the following distinction.
+
+For the purposes of versioning, all code within a release is treated as if it were part of a single package. All packages within a release receive the same version number. By keeping version numbers in lockstep, end users can understand which packages belong to which release.
```

Comment:
I understand this as every package getting a new version number when one package gets an update regardless if every package has had any changes or not.

This could introduce some issues. Semver 2.0 specifically says that increments are to be done to version numbers when incompatible API changes are made, functionality is introduced in a backwards compatible manner or bugs are fixed in a backwards compatibility manner. If a package has its version number increased without this happening, then the meaning of the version number is lost, since the user can't now tell if a package that went from X.Y.Z to X.Y.Z + 1 actually has a bug fix or not.

This also may cause problems when debugging issues. If version numbers are increased only when packages have changed then the user who is facing an issue can downgrade a package that may be causing this issue. If the issue persists, then the user knows that the changes in that package are not causing the issue. If version numbers are increased even if packages have no changes, now the user can't use version numbers for debugging purposes, because if the user downgrades a package and the issue persists, then the user can't tell if the issue is related or not because it is not possible to tell from the version number if the package has changed or not.

Reply:
> Semver 2.0 specifically says that increments are to be done to version numbers when incompatible API changes are made

That happens only when the major version changes. As long as we stick to minor/patch increases, we should be all fine.

> an user won't be able to tell if a package that went from X.Y.Z to X.Y.Z + 1 actually has a bug fix or not.

I'd argue CHANGELOG should be the reference for users in this case ;)

Reply:
Yes, we are saying, specifically, that semver applies to all the code in core, at once. Individual sub-packages are not separate code bases with separate version numbers. There is only one codebase, **opentelemetry-core**, which contains all the code for that language implementation. So users should look in the changelog when they update to the latest version.

Contrib packages, on the other hand, are stand-alone and can be versioned independently.

Reply:
@carlosalberto Actually Semver 2.0 specifically says that _increments are to be done to version numbers when incompatible API changes are made, functionality is introduced in a backwards compatible manner or bugs are fixed in a backwards compatibility manner_. What I mean with this is that increments are to be done to version numbers (major, minor and patch numbers) when code changes happen, so even if we only increase non-major version numbers it would be Semver 2.0 non-compliant to increase **any** number (major, minor or patch) if there is no change in the code associated with that version number increase.

Reply:
@tedsuo, if I understand correctly, the intention of keeping the version number in lockstep for all _core_ OpenTelemetry components, is to make it possible for users to understand which packages belong to which release. Now, maybe we are trying to tie the _OpenTelemetry release_ concept to the _OpenTelemetry individual package version number_ concept when they should be handled separately? As far as I understand, the latter concept has a meaning clearly defined already in Semver 2.0. The former concept may mean "a collection of specifically versioned individual OpenTelemetry individual packages"? :shrug:

I mean, maybe what we are trying to achieve here is to have another number, an _OpenTelemetry release_ number? For example, the OpenTelemetry Java release 7 means the collection of the OpenTelemetry Java API 3.5.1, OpenTelemetry Java SDK 4.5.5, and OpenTelemetry Java SomethingElse 8.1.3 packages?

We can give some meaning to the release number by saying _**this** set of tests was run using **this** release_ (which means, with these specific individual packages) so that we are 110% sure that everything works and is absolutely perfect and completely and absolutely bug-free and the bestest software ever (:stuck_out_tongue_closed_eyes: )?

Reply:
Also, I have some doubts regarding expecting users to rely on the CHANGELOG to find out if there were actually changes made in the individual packages. I would be a bit annoyed if I had to read CHANGELOG after CHANGELOG file to find out which was the last package that included a change when I am debugging something in the scenario described in my first comment and I would be asking myself if some tool that relies on version numbers to automate some task will work with OpenTelemetry.

Reply:
@ocelotl Do you propose "Release Trains"?

Reply:
> This could introduce some issues. Semver 2.0 specifically says that increments are to be done to version numbers when incompatible API changes are made, functionality is introduced in a backwards compatible manner or bugs are fixed in a backwards compatibility manner.

I feel we are running in circles. Simply: for 1.x versions, changes are expected to be have additive changes, and we won't introduce breaking changes.

> Also, I have some doubts regarding expecting users to rely on the CHANGELOG to find out if there were actually changes made in the individual packages. I would be a bit annoyed if I had to read CHANGELOG after CHANGELOG 

That's not different that when you upgrade to a new Java/DotNet versions, as you need to read the documentation (CHANGELOG may not be the most friendly one, but still). 

> to find out which was the last package that included a change when I am debugging something in the scenario described in my first comment 

You are trying to detect a bug based on the fact changes happened in a dependency? What if 15 packages included changes, but they are unrelated or code bases clean up? Do you still check them? If this is not the case, please elaborate.

Reply:
@iNikem Not necessarily :sweat_smile: I am simply suggesting that we could group individual packages and assign a number to this group so that the versioning numbers of individual packages can keep their usual meaning. Release Trains includes a lot more than just this, like scheduling, commitments, specific roles for participants, etc.

Reply:
@carlosalberto hmm, maybe I am not being clear, sorry. I just mean that it is not Semver 2.0 compliant to increase a version number if there are no changes associated with that increase. I think the root of this issue/misunderstanding is the concept of OpenTelemetry release. @tedsuo, what do you mean with _release_ in this statement:

> By keeping version numbers in lockstep, end users can understand which packages belong to which release.

?

Reply:
A release is a tested version of opentelemetry core. It is a tagged version of the core repo, and the resulting artifacts generated by that CI pipeline. Contrib packages are separate, but required to be compatible with the latest release.

> By keeping version numbers in lockstep, end users can understand which packages belong to which release.

By this, I mean that if users are bumping versions of individual subpackages, rather than bumping the entire core, then they can end up running an untested combination of packages. If we version every subpackage separately, then it is much harder for users to understand when they are running an official release, and when they are using an untested combination. OpenTelemetry is easier to comprehend when it is treated as a monolithic codebase.

But, based on many of these conversations, I'm coming around to the idea that the API and SDK may benefit from versioning separately, and the SDK marks breaking changes with major version bumps. This separation will make life difficult for maintainers. In most languages, it will probably result in splitting core into two repos. It has been simpler for *everyone* to treat OpenTelemetry as a single monolithic codebase. That said, I can see how in the long term it would be helpful to have an API version number and an SDK version number. This would allow the SDK to follow semver, and allow  breaking changes without implying the instrumentation API has changed.

But every subpackage of the SDK definitely should have the same version number. The SDK should be treated as a monolithic codebase in this regard. Experimental packages may be isolated and kept at v0 until they are stable. But  stable packages should be versioned together.

Note: there may be packages in core which should really be in contrib, such as exporters. For example, It may be reasonable for the prometheus exporter to keep its own version number, in which case it should be moved out of the SDK. But if it needs to be in the core SDK, it's part of the monolith.

Reply:
On this "helpful to have an API version number and an SDK version number." I think it's not just _"helpful"_, but **required**, if we assume the API surface is (long term) stable it doesn't make sense to rev the API version when there are no changes, it will just complicate the product development and potentially cause some products to just not upgrade because of potential breaking changes (assuming they don't read the Changelog and it's not explicitly called out that nothing actually changed).

So as you point out "This would allow the SDK to follow semver, and allow breaking changes without implying the instrumentation API has changed." I think this should be explicitly called out.

Reply:
I'd like to strongly advocate that we allow the SDK and API version numbers to differ. Once stable, the API should have few if any changes. SDKs are likely to undergo a number of bugfixes, additions, improvements, etc. Revving the API version to match the SDK version when there have not been any changes to the API is a very confusing signal. We hope and expect to have many third parties depending on the OpenTelemetry API over time. This potentially includes framework authors who are building in OpenTelemetry instrumentation hooks and alternative SDK implementors. Coupling the API version with the OpenTelemetry SDK could be challenging for third parties that depend on only the API.

Reply:
Strong :+1: to @mwear suggestion to keep the API version separate. We discussed this at length in the JS SIG today and I fully agree that keeping the API as unchanging as possible makes it significantly easier for package maintainers to build in support for opentelemetry, which is the goal.

Reply:
I am very strongly in favor of separate versions for SDK and API. I am fairly certain we had this discussion & conclusion almost a year ago, and are just rehashing those arguments.

Reply:
Great. Glad we're coming to consensus.